riomucho/__init__.py

```diff
@@ -1,95 +1,146 @@
+"""TODO
+"""
+
 from __future__ import with_statement
+from functools import wraps
 from multiprocessing import Pool
+import sys
+import traceback
+
+import click
+import numpy as np
 import rasterio as rio
 from rasterio.transform import guard_transform
-import numpy as np
-import click
+
 import riomucho.scripts.riomucho_utils as utils
 from riomucho.single_process_pool import MockTub
-import traceback
+
 
 work_func = None
 global_args = None
 srcs = None
 
 
+class MuchoChildError(Exception):
+    """A wrapper for exceptions in a child process.
+
+    See https://bugs.python.org/issue13831
+    """
+    def __init__(self):
+        """Wrap the last exception."""
+        exc_type, exc_value, exc_tb = sys.exc_info()
+        self.exception = exc_value
+        self.formatted = ''.join(traceback.format_exception(exc_type, exc_value, exc_tb))
+
+    def __str__(self):
+        return "{}\nChild process's traceback:\n{}".format(
+            Exception.__str__(self), self.formatted)
+
+
+def job(func):
+    """A decorator which captures job state.
+
+    Tracebacks in particular, are captured. Inspired by an example in
+    https://bugs.python.org/issue13831.
+
+    This decorator wraps rio-mucho jobs.
+
+    Parameters
+    ----------
+    func : function
+        A function to be decorated.
+
+    Returns
+    -------
+    func
+
+    """
+    @wraps(func)
+    def wrapper(*args, **kwds):
+        try:
+            return func(*args, **kwds)
+        except Exception:
+            raise MuchoChildError()
+
+    return wrapper
+
+
 def main_worker(inpaths, g_work_func, g_args):
-    """"""
+    """TODO"""
     global work_func
     global global_args
     global srcs
     work_func = g_work_func
     global_args = g_args
-    try:
-        srcs = [rio.open(i) for i in inpaths]
-    except:
-        return
+    srcs = [rio.open(i) for i in inpaths]
+
 
+@job
```

Comment:
The `job` decorator wraps up tracebacks that happen in child processes and reraises them, which is better than printing them because they can be caught and inspected in Python.